### ISSUE TITLE
-d:lto + non orc/arc mm + some machines fail uncompressing

### ISSUE THREAD

**Issue by guzba** [guzba/zippy#45]
https://github.com/zedeus/nitter/commit/d407051b66c3a7d0ed6a9aed16ad1a844af44398

**Comment by Yardanico** [comment]
Hi, we've discussed this issue in Nim Telegram chat and it really seems like zippy is doing undefined behaviour by disregarding strict aliasing/type punning rules in C.

You can reproduce that zedeus' issue on latest GCC with LTO for example (I can easily repro it with GCC + LTO on Arch in WSL).

The proper fix would be to use `memcpy` for all the read/write functions, with this change everything works fine:
```nim
proc c_memcpy(a, b: pointer, size: csize_t): pointer {.importc: "memcpy", header: "<string.h>", discardable.}

proc read16*(src: ptr UncheckedArray[uint8], ip: int): uint16 {.inline.} =
  discard c_memcpy(result.addr, cast[pointer](cast[int](src) + ip), 2)

proc read32*(src: ptr UncheckedArray[uint8], ip: int): uint32 {.inline.} =
  discard c_memcpy(result.addr, cast[pointer](cast[int](src) + ip), 4)

proc read64*(src: ptr UncheckedArray[uint8], ip: int): uint64 {.inline.} =
  discard c_memcpy(result.addr, cast[pointer](cast[int](src) + ip), 8)

proc write64*(dst: ptr UncheckedArray[uint8], op: int, v: uint64) {.inline.} =
  discard c_memcpy(cast[pointer](cast[int](dst) + op), v.addr, 8)

proc copy64*(dst, src: ptr UncheckedArray[uint8], op, ip: int) {.inline.} =
  write64(dst, op, read64(src, ip))

proc read16*(s: string, pos: int): uint16 {.inline.} =
  discard c_memcpy(result.addr, s[pos].addr, 2)

proc read32*(s: string, pos: int): uint32 {.inline.} =
  discard c_memcpy(result.addr, s[pos].addr, 4)
```

**Comment by Yardanico** [comment]
And yeah, you can just use `copyMem` instead :)

**Comment by guzba** [comment]
I need to be able to reproduce this issue before I do anything. I currently cannot reproduce this locally or with GitHub actions. It sounds like you are able to reproduce the issue with Arch in WSL? I'll give that a try.

**Comment by Yardanico** [comment]
@guzba you can try it on a native Arch installation, with Arch in WSL, or maybe on some other distro that has GCC 12.

**Comment by guzba** [comment]
I'm using ubuntu-latest. If "latest" doesn't have this, is GCC 12.0 like beta or new or something? Oh it literally just came out.

**Comment by Yardanico** [comment]
@guzba "ubuntu-latest" in GitHub Actions is actually Ubuntu 20.04 - https://github.com/guzba/zippy/runs/6784731132?check_suite_focus=true#step:1:4 . I'm not sure if Ubuntu 22.04 has GCC 12 or not, but yes, GCC 12 is new but it's already released. I haven't tested, it might be the case that GCC 11 also has this bug.

**Comment by guzba** [comment]
It may, I will need to check. I haven't tested newer GCC though so from my perspective it is unsupported until I do.

I tried ArchWSL and first run after install it failed with some glibc error so I'm not going that route. I'm just going to set up an actual VM for testing.

**Comment by guzba** [comment]
Hm, turns out I can't use choosenim on Ubuntu 22.04 where I have GCC 11: https://github.com/dom96/choosenim/issues/297 Fortunately building Nim 1.6.6 from source worked without any trouble. Nimble doesn't work though `could not import: SSL_get_peer_certificate`. I'll confirm this a second time and then open an issue over there.

I am able to reproduce this LTO issue with GCC 11 on Ubuntu 22.04. Easy reproducibility of the issue is huge so this is great.

**Comment by guzba** [comment]
Ok I've confirmed this on my test VM along with all the usual tests and GitHub Actions. I've tagged a release including this fix https://github.com/guzba/zippy/releases/tag/0.9.11

Thanks for the fix Yardanico. I was vaguely aware of the strict aliasing rules and I'm somewhat disturbed I need to pull out the sledge hammer that is copyMem to just read a few bytes but if GCC needs that to not make some optimizations, it's a small price to pay.
